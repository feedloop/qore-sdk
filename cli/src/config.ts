import Conf, { Schema } from "conf";

export type CLIConfig = { token: string; project: string; org: string };

export const schema: Schema<CLIConfig> = {
  token: { type: "string", description: "Token" },
  project: { type: "string", description: "Project ID" },
  org: { type: "string", description: "Organization ID" }
};

const config = new Conf({
  schema
});

export default config;
