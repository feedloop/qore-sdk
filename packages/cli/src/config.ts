import Conf, { Schema } from "conf";

export type CLIConfig = { adminSecret: string; url: string };

export const schema: Schema<CLIConfig> = {
  adminSecret: {
    type: "string",
    description: "admin secret",
    default: "admin-secret"
  },
  url: {
    type: "string",
    description: "project url",
    default: "http://localhost:8080/"
  }
};

const config = new Conf({
  schema
});

export default config;
