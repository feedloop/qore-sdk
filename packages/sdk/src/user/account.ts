import { callApi } from "../common";
import { UrlUserPath } from "./url";
import { APIUser } from "./user";

export type APIAccount = {
  id: string;
  user: APIUser;
  type: string;
};

export type Account = APIAccount & {
  delete(): Promise<void>;
  updateType(type: string): Promise<void>;
};

export class AccountImpl implements Account {
  id: string;
  user: APIUser;
  type: string;
  _orgId: string;
  _url: UrlUserPath;
  _token: string;
  constructor(
    params: APIAccount & { url: UrlUserPath; userToken: string; orgId: string }
  ) {
    this.id = params.id;
    this.user = params.user;
    this.type = params.type;
    this._orgId = params.orgId;
    this._url = params.url;
    this._token = params.userToken;
  }
  async updateType(type: string) {
    await callApi(
      {
        method: "patch",
        url: this._url.account(this._orgId, this.id),
        data: { type }
      },
      this._token
    );
  }
  async delete() {
    await callApi(
      {
        method: "delete",
        url: this._url.account(this._orgId, this.id)
      },
      this._token
    );
  }
}
