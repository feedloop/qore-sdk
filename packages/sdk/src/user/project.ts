import { callApi } from "../common";
import { UrlUserPath } from "./url";

export type APIProject = {
  id: string;
  name: string;
};
export type Table = {
  id: string;
  name: string;
};

export type Project = APIProject & {
  tables(props?: { limit: number; offset: number }): Promise<Table[]>;
  delete(): Promise<void>;
  update(project: Partial<APIProject>): Promise<void>;
};

export class ProjectImpl implements Project {
  id: string;
  name: string;
  _orgId: string;
  _url: UrlUserPath;
  _token: string;
  constructor(
    params: APIProject & { url: UrlUserPath; userToken: string; orgId: string }
  ) {
    this.id = params.id;
    this.name = params.name;
    this._orgId = params.orgId;
    this._url = params.url;
    this._token = params.userToken;
  }
  async get() {
    await callApi(
      {
        method: "get",
        url: this._url.project(this._orgId)
      },
      this._token
    );
  }
  async tables(
    props: { limit?: number; offset?: number } = {}
  ): Promise<Table[]> {
    const { nodes } = await callApi(
      {
        method: "get",
        url: this._url.project(this._orgId, this.id) + "/tables"
      },
      this._token
    );
    return nodes;
  }
  async delete() {
    await callApi(
      {
        method: "delete",
        url: this._url.project(this._orgId, this.id)
      },
      this._token
    );
  }
  async update(project: Partial<APIProject>) {
    await callApi(
      {
        method: "patch",
        url: this._url.project(this._orgId, this.id),
        data: project
      },
      this._token
    );
  }
}
