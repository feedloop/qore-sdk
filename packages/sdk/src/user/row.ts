export type Rows<T = Record<string, any>> = {
  nodes: T[];
  totalCount: number;
};
