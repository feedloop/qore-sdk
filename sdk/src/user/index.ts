export { default } from "./user";
export * from "./user";
export * from "./account";
export * from "./organization";
export * from "./project";
export * from "./row";
export * from "./url";
