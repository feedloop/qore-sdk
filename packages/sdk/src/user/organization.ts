import { callApi } from "../common";
import { QoreProjectSchema } from "../project";
import { AccountImpl, APIAccount, Account } from "./account";
import { APIProject, Project, ProjectImpl } from "./project";
import { Rows } from "./row";
import { UrlUserPath } from "./url";

export type APIOrganization = {
  id: string;
  name: string;
  category: string;
  size: string;
  subdomain: string;
};

export type Organization = APIOrganization & {
  accounts(limit?: number, offset?: number): Promise<Account[]>;
  inviteAccount(params: { email: string; type: string }): Promise<void>;
  createProject(params: {
    name: string;
    schema?: QoreProjectSchema;
  }): Promise<string>;
  projects(
    limit?: number,
    offset?: number
  ): Promise<{ projects: Project[]; totalCount: number }>;
  project(id: string): Promise<Project>;
  delete(): Promise<void>;
  update(org: Partial<APIOrganization>): Promise<void>;
};

export class OrganizationImpl implements Organization {
  id: string;
  name: string;
  category: string;
  subdomain: string;
  size: string;
  _url: UrlUserPath;
  _token: string;
  constructor(
    params: APIOrganization & { url: UrlUserPath; userToken: string }
  ) {
    this.id = params.id;
    this.name = params.name;
    this.subdomain = params.subdomain;
    this.category = params.category;
    this.size = params.size;
    this._url = params.url;
    this._token = params.userToken;
  }
  async accounts(limit?: number, offset?: number): Promise<Account[]> {
    const { nodes } = await callApi<Rows<APIAccount>>(
      {
        method: "get",
        url: this._url.account(this.id),
        params: { limit, offset }
      },
      this._token
    );
    return nodes.map(
      account =>
        new AccountImpl({
          ...account,
          userToken: this._token,
          url: this._url,
          orgId: this.id
        })
    );
  }
  async inviteAccount(params: { email: string; type: string }): Promise<void> {
    await callApi(
      {
        method: "post",
        url: this._url.account(this.id),
        data: params
      },
      this._token
    );
  }
  async createProject(params: {
    name: string;
    schema?: QoreProjectSchema;
  }): Promise<string> {
    const { id } = await callApi(
      {
        method: "post",
        url: this._url.project(this.id),
        data: params
      },
      this._token
    );
    return id;
  }
  async projects(
    limit = 10,
    offset = 0
  ): Promise<{ projects: Project[]; totalCount: number }> {
    const { nodes, totalCount } = await callApi(
      {
        method: "get",
        url: this._url.project(this.id),
        params: { limit, offset }
      },
      this._token
    );
    return {
      projects: nodes.map(
        (row: APIProject) =>
          new ProjectImpl({
            ...row,
            userToken: this._token,
            url: this._url,
            orgId: this.id
          })
      ),
      totalCount
    };
  }
  async project(id: string): Promise<Project> {
    const project = await callApi<APIProject>(
      {
        method: "get",
        url: this._url.project(this.id, id)
      },
      this._token
    );
    return new ProjectImpl({
      ...project,
      userToken: this._token,
      url: this._url,
      orgId: this.id
    });
  }
  async delete(): Promise<void> {
    await callApi(
      {
        method: "delete",
        url: this._url.organization(this.id)
      },
      this._token
    );
  }
  async update(org: Partial<APIOrganization>): Promise<void> {
    await callApi(
      {
        method: "patch",
        url: this._url.organization(this.id),
        data: org
      },
      this._token
    );
  }
}
