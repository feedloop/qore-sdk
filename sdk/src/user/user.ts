import { callApi } from '../common';
import { Project, ProjectImpl } from './project';
import { APIOrganization, Organization, OrganizationImpl } from './organization';
import { generateUrlUserPath } from './url';

export type APIUser = {
  email: string;
  token: string;
};

export type QoreUser = {
  register(params: { email: string; password: string }): Promise<void>;
  logout(): void;
  login(email: string, password: string): Promise<void>;
  verify(email: string, activationCode: string): Promise<void>;
  createOrganization(params: { name: string; category: string; size: string }): Promise<string>;
  organizations(limit?: number, offset?: number): Promise<Organization[]>;
  organization(id: string): Promise<Organization | undefined>;
  projects(orgId: string, props?: { limit?: number; offset?: number }): Promise<Project[]>;
  user(): APIUser | undefined;
};

export default () => {
  let user: APIUser;
  const url = generateUrlUserPath();
  return {
    setToken(token: string) {
      user = { email: '', token };
    },
    async register(params: { email: string; password: string }): Promise<void> {
      await callApi({
        method: 'post',
        url: url.register(),
        data: params,
      });
    },
    async login(email: string, password: string): Promise<void> {
      const { token } = await callApi({
        method: 'post',
        url: url.login(),
        data: { email, password },
      });
      user = { email, token: 'Bearer ' + token };
    },
    async verify(email: string, activationCode: string): Promise<void> {
      const { token } = await callApi({
        method: 'post',
        url: url.verify(),
        data: { email, activationCode },
      });
      user = { email, token: 'Bearer ' + token };
    },
    async createOrganization(params: {
      name: string;
      category: string;
      size: string;
      subdomain: string;
    }): Promise<string> {
      const { id } = await callApi(
        {
          method: 'post',
          url: url.organization(),
          data: params,
        },
        user.token
      );
      return id;
    },
    async projects(orgId: string, props: { limit?: number; offset?: number }): Promise<Project[]> {
      const { limit, offset } = props || {};
      const { nodes } = await callApi(
        {
          method: 'get',
          url: url.project(orgId),
          params: { limit, offset },
        },
        user.token
      );
      return nodes.map(
        (row: APIOrganization) => new ProjectImpl({ ...row, userToken: user.token, url, orgId })
      );
    },
    async organizations(limit?: number, offset?: number): Promise<Organization[]> {
      const { nodes } = await callApi(
        {
          method: 'get',
          url: url.organization(),
          params: { limit, offset },
        },
        user.token
      );
      return nodes.map(
        (row: APIOrganization) => new OrganizationImpl({ ...row, userToken: user.token, url })
      );
    },
    async organization(id: string): Promise<Organization> {
      // NOT IMPLEMENTED IN BACKEND
      // const org = await callApi<APIOrganization>(
      //   {
      //     method: 'get',
      //     url: url.organization(id),
      //   },
      //   user.token
      // );
      return new OrganizationImpl({
        ...{ id, category: '', size: '', name: '', subdomain: '' },
        userToken: user.token,
        url,
      });
    },
    user: () => user,
    logout: () => {
      user = { token: '', email: '' };
    },
  };
};
