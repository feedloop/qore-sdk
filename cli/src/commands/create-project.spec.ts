import { setupRecorder } from "nock-record";
import fse from "fs-extra";
import { default as makeUser } from "@feedloop/qore-sdk/lib/user";
import path from "path";
import CreateProject from "./create-project";
import config from "../config";

const record = setupRecorder();
const projectName = "qore-project-test";

describe("create-project", () => {
  beforeEach(() => {
    config.reset("org", "project", "token");
  });
  afterAll(() => {
    fse.removeSync(path.resolve(process.cwd(), projectName));
  });
  it("should be able to export schema", async () => {
    const { completeRecording } = await record("create-project");
    await CreateProject.run([
      "--org",
      "lIdfC42DJCN2XzQ",
      "--token",
      "3960f3b8-a139-42eb-8295-3d669e4da4c9",
      projectName
    ]);

    const qoreConfig = await fse.readJson(
      path.resolve(process.cwd(), projectName, "qore.config.json")
    );
    expect(qoreConfig).toMatchSnapshot();
    await expect(
      CreateProject.run([
        "--org",
        "lIdfC42DJCN2XzQ",
        "--token",
        "3960f3b8-a139-42eb-8295-3d669e4da4c9",
        "--template",
        "some-unknown-template",
        projectName
      ])
    ).rejects.toThrowError(
      'Cant find "some-unknown-template" from project templates, may want to choose from the following available templates: todo-list-typescript'
    );
    completeRecording();
  });
});
