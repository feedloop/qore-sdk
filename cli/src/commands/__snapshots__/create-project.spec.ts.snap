// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-project should be able to export schema 1`] = `
"{
  \\"version\\": \\"v1\\",
  \\"endpoint\\": \\"https://qore-api.feedloop.io\\",
  \\"project\\": \\"some-project-id\\",
  \\"org\\": \\"some-org-id\\"
}
"
`;
